source/particle/world.cc

```diff
@@ -1285,32 +1285,71 @@ namespace aspect
 
       const FiniteElement<dim> &velocity_fe = this->get_fe().base_element(this->introspection().base_elements.velocities);
 
+      const bool compute_fluid_velocity = this->include_melt_transport() && property_manager->get_data_info().fieldname_exists("melt_particle");
+      unsigned int fluid_component_index;
+      std::vector<bool> use_solid_velocity;
+      if (compute_fluid_velocity)
+        {
+          fluid_component_index = this->introspection().variable("fluid velocity").first_component_index;
+          const unsigned int melt_property_index = property_manager->get_property_component_by_name("melt_particle");
+
+          use_solid_velocity.resize(particles_in_cell,true);
+
+          typename std::multimap<types::LevelInd, Particle<dim> >::iterator it = begin_particle;
+          for (unsigned int particle_index = 0; it!=end_particle; ++it,++particle_index)
+            if (*it->get_properties()[melt_property_index] == 1.0)
+              use_solid_velocity[particle_index] = false;
+        }
+
       for (unsigned int j=0; j<velocity_fe.dofs_per_cell; ++j)
         {
-          Tensor<1,dim> solution_at_support_point;
-          Tensor<1,dim> old_solution_at_support_point;
+          Tensor<1,dim> velocity_at_support_point;
+          Tensor<1,dim> old_velocity_at_support_point;
+
+          Tensor<1,dim> fluid_velocity_at_support_point;
+          Tensor<1,dim> old_fluid_velocity_at_support_point;
+
           for (unsigned int dir=0; dir<dim; ++dir)
             {
               const unsigned int support_point_index
                 = this->get_fe().component_to_system_index(/*velocity component=*/ this->introspection().component_indices.velocities[dir],
                                                                                    /*dof index within component=*/ j);
-              solution_at_support_point[dir] = this->get_solution()[cell_dof_indices[support_point_index]];
-              old_solution_at_support_point[dir] = this->get_old_solution()[cell_dof_indices[support_point_index]];
+              velocity_at_support_point[dir] = this->get_solution()[cell_dof_indices[support_point_index]];
+              old_velocity_at_support_point[dir] = this->get_old_solution()[cell_dof_indices[support_point_index]];
```

Comment:
if I see this right below, then you don't actually need these two velocities if you want to use the melt velocity. maybe put their initialization into an `else` block of the following `if`?